File: ACR122U NFC Reader SDK/Sample Codes/MS Visual Basic.NET 2005/Device Programming/Basic Device Programming/Basic DeviceProgramming/Form1.vb
Public Class DeviceProgramming

    Dim retCode, hContext, hCard, Protocol As Long
    Dim connActive As Boolean = False
    Dim SendBuff(262), RecvBuff(262) As Byte
    Dim SendLen, RecvLen, nBytesRet As Integer
    Dim sReaderList As String
    Dim sReaderGroup As String
    Dim ioRequest As SCARD_IO_REQUEST


    Private Sub DeviceProgramming_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Call InitMenu()

    End Sub

    Private Sub InitMenu()

        connActive = False
        cbReader.Text = ""
        mMsg.Text = ""
        mMSg.Clear()
        bInit.Enabled = True
        bConnect.Enabled = False    
        bClear.Enabled = False
        gbAntenna.Enabled = False
        gbRed.Enabled = False
        gbGreen.Enabled = False
        gbBlinkDuration.Enabled = False
        bSetLED.Enabled = False
        bGetFW.Enabled = False
        Call displayOut(0, 0, "Program ready")


    End Sub

    Private Sub EnableButtons()

        bInit.Enabled = False
        bConnect.Enabled = True
        bClear.Enabled = True
        
    End Sub

    Private Sub ClearBuffers()

        Dim indx As Long

        For indx = 0 To 262
            RecvBuff(indx) = &H0
            SendBuff(indx) = &H0
        Next indx

    End Sub


    Private Sub displayOut(ByVal errType As Integer, ByVal retVal As Integer, ByVal PrintText As String)

        Select Case errType

            Case 0
                mMSg.SelectionColor = Drawing.Color.Green
            Case 1

                PrintText = ModWinsCard.GetScardErrMsg(retVal)
                mMSg.SelectionColor = Drawing.Color.Red
            Case 2
                mMSg.SelectionColor = Drawing.Color.Black
                PrintText = "<" + PrintText
            Case 3
                mMSg.SelectionColor = Drawing.Color.Black
                PrintText = ">" + PrintText
            Case 4
                mMSg.SelectionColor = Drawing.Color.Green

        End Select

        mMSg.SelectedText = PrintText & vbCrLf
        mMSg.SelectionColor = Drawing.Color.Black
        mMSg.SelectionStart = mMSg.TextLength
        mMsg.Focus()

    End Sub

    Private Sub bInit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bInit.Click


        Dim ReaderCount As Integer
        Dim ctr As Integer


        For ctr = 0 To 255
            sReaderList = sReaderList + vbNullChar
        Next

        ReaderCount = 255

        ' 1. Establish context and obtain hContext handle
        retCode = ModWinsCard.SCardEstablishContext(ModWinsCard.SCARD_SCOPE_USER, 0, 0, hContext)

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then

            Call displayOut(1, retCode, "")

            Exit Sub

        End If

        ' 2. List PC/SC card readers installed in the system
        retCode = ModWinsCard.SCardListReaders(hContext, "", sReaderList, ReaderCount)

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then

            Call displayOut(1, retCode, "")

            Exit Sub

        End If

        ' Load Available Readers
        Call LoadListToControl(cbReader, sReaderList)
        cbReader.SelectedIndex = 0
        Call EnableButtons()

    End Sub


    Public Sub LoadListToControl(ByVal Ctrl As ComboBox, ByVal ReaderList As String)

        Dim sTemp As String
        Dim indx As Integer

        indx = 1
        sTemp = ""
        Ctrl.Items.Clear()

        While (Mid(ReaderList, indx, 1) <> vbNullChar)

            While (Mid(ReaderList, indx, 1) <> vbNullChar)
                sTemp = sTemp + Mid(ReaderList, indx, 1)
                indx = indx + 1
            End While

            indx = indx + 1

            Ctrl.Items.Add(sTemp)

            sTemp = ""
        End While

    End Sub

    Private Sub bConnect_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bConnect.Click


        ' Connect to selected reader using hContext handle and obtain hCard handle
        If connActive Then

            retCode = ModWinsCard.SCardDisconnect(hCard, ModWinsCard.SCARD_UNPOWER_CARD)

        End If

        ' Shared Connection
        retCode = ModWinsCard.SCardConnect(hContext, cbReader.SelectedItem.ToString(), ModWinsCard.SCARD_SHARE_SHARED, ModWinsCard.SCARD_PROTOCOL_T0 Or ModWinsCard.SCARD_PROTOCOL_T1, hCard, Protocol)

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then
            Exit Sub


        Else

            Call displayOut(0, 0, "Successful connection to " & cbReader.Text)

        End If

        connActive = True
        bGetFW.Enabled = True
        gbAntenna.Enabled = True
        gbRed.Enabled = True
        gbGreen.Enabled = True
        gbBlinkDuration.Enabled = True
        bSetLED.Enabled = True
        bGetFW.Enabled = True
        rbAntOn.Checked = True
        rbRedFinalOff.Checked = True
        rbRedStateMaskOff.Checked = True
        rbRedInitOff.Checked = True
        rbRedBlinkMaskOff.Checked = True
        rbGreenFinalOff.Checked = True
        rbGreenStateMaskOff.Checked = True
        rbGreenInitOff.Checked = True
        rbGreenBlinkMaskOff.Checked = True
        rbLinkToBuzzOpt1.Checked = True
        tT1.Text = "00"
        tT2.Text = "00"
        tRepeat.Text = "01"

    End Sub

    Private Sub bGetFW_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bGetFW.Click

        ' Get the firmaware version of the reader
        Dim tmpStr As String
        Dim intIndx As Integer

        Call ClearBuffers()
        SendBuff(0) = &HFF
        SendBuff(1) = &H0
        SendBuff(2) = &H48
        SendBuff(3) = &H0
        SendBuff(4) = &H0
        SendLen = 5
        RecvLen = 10

        retCode = Transmit

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then

            Exit Sub

        End If


        ' Interpret firmware data
        tmpStr = "Firmware Version: "

        For intIndx = 0 To RecvLen

            If RecvBuff(intIndx) <> &H0 Then

                tmpStr = tmpStr & Chr(RecvBuff(intIndx))

            End If

        Next intIndx

        Call displayOut(3, 0, tmpStr)

    End Sub

    Private Function Transmit() As Long

        Dim tmpStr As String
        Dim indx As Integer

        'Display Apdu In
        tmpStr = ""

        For indx = 0 To SendLen - 1

            tmpStr = tmpStr + Microsoft.VisualBasic.Right("00" & Hex(SendBuff(indx)), 2) + " "

        Next indx

        Call DisplayOut(2, 0, tmpStr)

        ioRequest.dwProtocol = Protocol
        ioRequest.cbPciLength = Len(ioRequest)

        RecvLen = 262

        'Issue SCardTransmit
        retCode = ModWinsCard.SCardTransmit(hCard, _
                             ioRequest, _
                             SendBuff(0), _
                             SendLen, _
                             ioRequest, _
                             RecvBuff(0), _
                             RecvLen)

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then

            Call displayOut(1, retCode, "")

        Else

            tmpStr = ""

            For indx = 0 To (RecvLen - 1)

                tmpStr = tmpStr + Microsoft.VisualBasic.Right("00" & Hex(RecvBuff(indx)), 2) + " "

            Next indx

            Call displayOut(3, 0, Trim(tmpStr))

        End If

        Transmit = retCode

    End Function

    Private Sub bClear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bClear.Click

        mMsg.Clear()

    End Sub

    Private Sub bReset_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bReset.Click

        If connActive Then

            retCode = ModWinsCard.SCardDisconnect(hCard, ModWinsCard.SCARD_UNPOWER_CARD)

        End If

        retCode = ModWinsCard.SCardReleaseContext(hCard)
        Call InitMenu()

    End Sub

    Private Sub bQuit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bQuit.Click

        ' terminate the application
        retCode = ModWinsCard.SCardReleaseContext(hContext)
        retCode = ModWinsCard.SCardDisconnect(hCard, ModWinsCard.SCARD_UNPOWER_CARD)
        End

    End Sub

    Private Sub bSetAntenna_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bSetAntenna.Click

        'set the antenna options
        Call ClearBuffers()
        SendBuff(0) = &HFF
        SendBuff(1) = &H0
        SendBuff(2) = &H0
        SendBuff(3) = &H0
        SendBuff(4) = &H4
        SendBuff(5) = &HD4
        SendBuff(6) = &H32
        SendBuff(7) = &H1

        If rbAntOn.Checked = True Then

            SendBuff(8) = &H1

        Else

            SendBuff(8) = &H0

        End If

        SendLen = 9

        retCode = Transmit()

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then

            Exit Sub

        End If

    End Sub

    Private Sub bSetLED_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles bSetLED.Click

        ' Set the LED/Buzzer Settings
        Dim tmpStr As String
        Dim intIndx As Long
        Dim tmpLong As Byte

        ' Validate input
        If (tT1.Text = "" Or Not Byte.TryParse(tT1.Text.Trim(), System.Globalization.NumberStyles.HexNumber, Nothing, tmpLong)) Then

            tT1.Focus()
            tT1.Text = ""
            Exit Sub

        End If

        If (tT2.Text = "" Or Not Byte.TryParse(tT2.Text.Trim(), System.Globalization.NumberStyles.HexNumber, Nothing, tmpLong)) Then

            tT2.Focus()
            tT2.Text = ""
            Exit Sub

        End If

        If (tRepeat.Text = "" Or Not Byte.TryParse(tRepeat.Text.Trim(), System.Globalization.NumberStyles.HexNumber, Nothing, tmpLong)) Then

            tRepeat.Focus()
            tRepeat.Text = ""
            Exit Sub

        End If

        If tT1.Text = "" Then

            tT1.Focus()
            Exit Sub

        End If


        If tT2.Text = "" Then

            tT2.Focus()
            Exit Sub

        End If

        If tRepeat.Text = "" Then

            tRepeat.Focus()
            Exit Sub

        End If

        Call ClearBuffers()
        SendBuff(0) = &HFF
        SendBuff(1) = &H0
        SendBuff(2) = &H40
        SendBuff(3) = &H0

        If rbRedFinalOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H1

        End If

        If rbGreenFinalOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H2

        End If

        If rbRedStateMaskOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H4

        End If

        If rbGreenStateMaskOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H8

        End If

        If rbRedInitOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H10

        End If

        If rbGreenInitOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H20

        End If

        If rbRedBlinkMaskOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H40

        End If

        If rbGreenBlinkMaskOn.Checked = True Then

            SendBuff(3) = SendBuff(3) Or &H80

        End If

        SendBuff(4) = &H40
        SendBuff(5) = CByte("&H" + tT1.Text)
        SendBuff(6) = CByte("&H" + tT2.Text)
        SendBuff(7) = CByte("&H" + tRepeat.Text)

        If rbLinkToBuzzOpt1.Checked = True Then

            SendBuff(8) = &H0

        End If

        If rbLinkToBuzzOpt2.Checked = True Then

            SendBuff(8) = &H1

        End If

        If rbLinkToBuzzOpt3.Checked = True Then

            SendBuff(8) = &H2

        End If

        If rbLinkToBuzzOpt4.Checked = True Then

            SendBuff(8) = &H3

        End If

        SendLen = 9
        RecvLen = 2

        retCode = Transmit()

        If retCode <> ModWinsCard.SCARD_S_SUCCESS Then

            Exit Sub

        End If

    End Sub

End Class

File: ACR122U NFC Reader SDK/Sample Codes/MS Visual C#.NET 2005/Device Programming/Peer to Peer Programming/Active Sample/Active Sample/Form1.cs
//  Copyright(C):      Advanced Card Systems Ltd
//
//  File:              frmActive.pas
//
//  Description:       This sample program outlines the steps
//                     on how to set an ACR122 NFC reader to its
//                     active mode and sending data
//
//  Author:            Wazer Emmanuel R. Benal
//
//  Date:              August 5, 2008
//
//  Revision Trail:   (Date/Author/Description)
//
//======================================================================

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace ActiveSample
{
    public partial class ActiveSample : Form
    {
        public int retCode, hContext, hCard, Protocol, ReaderCount, nBytesRet;
        public bool connActive = false;
        public byte[] SendBuff = new byte[263];
        public byte[] RecvBuff = new byte[263];
        public int SendLen, RecvLen, ReaderLen, ATRLen, dwState, dwActProtocol;
        public byte[] ATRVal = new byte[257];
        string sReaderList;
        byte[] sReaderGroup;
        ModWinsCard.SCARD_IO_REQUEST ioRequest;

        public ActiveSample()
        {
            InitializeComponent();
        }

        private void InitMenu()
        {
            connActive = false;
            cbReader.Text = "";
            mMsg.Text = "";
            bInit.Enabled = true;
            bConnect.Enabled = false;
            bClear.Enabled = false;
            displayOut(0, 0, "Program ready");
            bActive.Enabled = false;

        }

        private void EnableButtons()
        {

            bInit.Enabled = false;
            bConnect.Enabled = true;
            bClear.Enabled = true;
            bActive.Enabled = true;

        }

        private void ClearBuffers()
        {
            long indx;

            for (indx = 0; indx <= 262; indx++)
            {
                RecvBuff[indx] = 0;
                SendBuff[indx] = 0;
            }

        }    

        private void displayOut(int errType, int retVal, string PrintText)
        {

            switch (errType)
            {

                case 0:
                    mMsg.SelectionColor = Color.Green;
                    break;
                case 1:
                    mMsg.SelectionColor = Color.Red;
                    PrintText = ModWinsCard.GetScardErrMsg(retVal);
                    break;
                case 2:
                    mMsg.SelectionColor = Color.Black;
                    PrintText = "<" + PrintText;
                    break;
                case 3:
                    mMsg.SelectionColor = Color.Black;
                    PrintText = ">" + PrintText;
                    break;
                case 4:
                    break;

            }

            mMsg.AppendText(PrintText);
            mMsg.AppendText("\n");
            mMsg.SelectionColor = Color.Black;
            mMsg.Focus();

        }
        private void bInit_Click(object sender, EventArgs e)
        {

            string ReaderList = "" + Convert.ToChar(0);
            int indx;
            int pcchReaders = 0;
            string rName = "";

            //Establish Context
            retCode = ModWinsCard.SCardEstablishContext(ModWinsCard.SCARD_SCOPE_USER, 0, 0, ref hContext);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {

                displayOut(1, retCode, "");

                return;

            }

            // 2. List PC/SC card readers installed in the system

            retCode = ModWinsCard.SCardListReaders(this.hContext, null, null, ref pcchReaders);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {

                displayOut(1, retCode, "");

                return;
            }

            EnableButtons();

            byte[] ReadersList = new byte[pcchReaders];

            // Fill reader list
            retCode = ModWinsCard.SCardListReaders(this.hContext, null, ReadersList, ref pcchReaders);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                mMsg.AppendText("SCardListReaders Error: " + ModWinsCard.GetScardErrMsg(retCode));

                return;
            }
            else
            {
                displayOut(0, 0, " ");
            }

            rName = "";
            indx = 0;

            //Convert reader buffer to string
            while (ReadersList[indx] != 0)
            {

                while (ReadersList[indx] != 0)
                {
                    rName = rName + (char)ReadersList[indx];
                    indx = indx + 1;
                }

                //Add reader name to list
                cbReader.Items.Add(rName);
                rName = "";
                indx = indx + 1;

            }

            if (cbReader.Items.Count > 0)
            {
                cbReader.SelectedIndex = 0;

            }
        }

        private void bConnect_Click(object sender, EventArgs e)
        {
            retCode = ModWinsCard.SCardConnect(hContext, cbReader.SelectedItem.ToString(), ModWinsCard.SCARD_SHARE_SHARED,
                                           ModWinsCard.SCARD_PROTOCOL_T1, ref hCard, ref Protocol);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                retCode = ModWinsCard.SCardConnect(hContext, cbReader.SelectedItem.ToString(), ModWinsCard.SCARD_SHARE_DIRECT,
                                            0, ref hCard, ref Protocol);


                if (retCode != ModWinsCard.SCARD_S_SUCCESS)


                    displayOut(1, retCode, "");
                else
                {
                    displayOut(0, 0, "Successful connection to " + cbReader.Text);

                }
            }
            else
            {
                displayOut(0, 0, "Successful connection to " + cbReader.Text);
            }
            GetFirmware();
        
        }

        private void GetFirmware()
        {

            // Get the firmaware version of the reader
            string tmpStr;
            int intIndx;

            ClearBuffers();
            SendBuff[0] = 0xFF;
            SendBuff[1] = 0x00;
            SendBuff[2] = 0x48;
            SendBuff[3] = 0x00;
            SendBuff[4] = 0x00;
            SendLen = 5;
            RecvLen = 10;

            retCode = CardControl();

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {

                return;

            }


            // Interpret firmware data
            tmpStr = "Firmware Version: ";

            for (intIndx = 0; intIndx <= RecvLen; intIndx++)
            {

                if (RecvBuff[intIndx] != 0)
                {

                    tmpStr = tmpStr + System.Text.ASCIIEncoding.ASCII.GetString(RecvBuff, 0, 19);
                }

            }

            displayOut(3, 0, tmpStr);
        }


        private void bClear_Click(object sender, EventArgs e)
        {
            mMsg.Clear();
        }

        private void bReset_Click(object sender, EventArgs e)
        {
            if (connActive)
            {

                retCode = ModWinsCard.SCardDisconnect(hCard, ModWinsCard.SCARD_UNPOWER_CARD);

            }

            retCode = ModWinsCard.SCardReleaseContext(hCard);
            InitMenu();
        }

        private void ActiveSample_Load(object sender, EventArgs e)
        {
            InitMenu();
        }

        private void bQuit_Click(object sender, EventArgs e)
        {
            // terminate the application
            retCode = ModWinsCard.SCardReleaseContext(hContext);
            retCode = ModWinsCard.SCardDisconnect(hCard, ModWinsCard.SCARD_UNPOWER_CARD);
            System.Environment.Exit(0);

        }

        private void bActive_Click(object sender, EventArgs e)
        {
            SetActive();
        }

        public void SetActive()
        {
            int indx;
            String tempstr="";
            ClearBuffers();
            
            SendBuff[0] = 0xFF;
            SendBuff[1] = 0x00;
            SendBuff[2] = 0x00;
            SendBuff[3] = 0x00;
            SendBuff[4] = 0x0A;
            SendBuff[5] = 0xD4;
            SendBuff[6] = 0x56;
            SendBuff[7] = 0x01;
            SendBuff[8] = 0x02;
            SendBuff[9] = 0x01;
            SendBuff[10] = 0x00;
            SendBuff[11] = 0xFF;
            SendBuff[12] = 0xFF;
            SendBuff[13] = 0x00;
            SendBuff[14] = 0x00;
            
            SendLen = 15;
            RecvLen = 21;


            retCode = CardControl();
            if (retCode != ModWinsCard.SCARD_S_SUCCESS)            
                return;

      
            for (indx = RecvLen - 2; indx <= RecvLen - 1; indx++)
            {
                tempstr = tempstr + " " + string.Format("{0:X2}", RecvBuff[indx]);
            }

            if (tempstr.Trim() != "90 00")
            {
                displayOut(2, 0, "Set active failed!");
                return;
            }

            sending();
        
        }

        public int CardControl()
        {

            string tempstr = "";
            int index = 0;

            for (index = 0; index <= SendLen - 1; index++)
            {               

                tempstr = tempstr + " " + string.Format("{0:X2}", SendBuff[index]);
            }

            displayOut(2,0,tempstr);

            
            retCode = ModWinsCard.SCardControl(hCard, (int)ModWinsCard.IOCTL_CCID_ESCAPE_SCARD_CTL_CODE, ref SendBuff[0], SendLen, ref RecvBuff[0], RecvLen, ref nBytesRet);

            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                
                displayOut(2, retCode, "");
                return retCode;
            }

            tempstr = "";

            for (index = 0; index <= RecvLen - 1; index++)
            {
               
                tempstr = tempstr + " " + string.Format("{0:X2}", RecvBuff[index]);
            }

            //DisplayOut(tempstr, 4);
            displayOut(3, 0, tempstr);

            return retCode;

        }

        public void sending()
        {
            char[] tempdata = new char[513];
            byte[] tmpbyte;
            byte tmpLong;
            string data = string.Empty;
            int datalen;
            int index = 0;

            //Transfer string data to a char array
            //and determine its length
            data = tbData.Text;
            datalen = data.Length;

            for (index = 0; index <= datalen - 1; index++)
            {
          
                tempdata[index] = data[index];
            
            }

            if (tbData.Text == "" | !byte.TryParse(tbData.Text.Trim(), System.Globalization.NumberStyles.HexNumber, null, out tmpLong))
            {
                tbData.Focus();
                tbData.Text = "";
                return;
            }


            //Send the length of the data first
            //so that the receiving device would
            //know how long the data would be
            ClearBuffers();
            SendBuff[0] = 0xFF;
            SendBuff[1] = 0x00;
            SendBuff[2] = 0x00;
            SendBuff[3] = 0x00;
            SendBuff[4] = 0x01;
            SendBuff[5] = 0xD4;
            SendBuff[6] = 0x40;
            SendBuff[7] = 0x01;       
            SendBuff[8] = (byte)datalen;

            //the length of the data

            SendLen = 9;
            RecvLen = 7;

            retCode = CardControl();
            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                return;
            }

            //Send actual data
            ClearBuffers();
            SendBuff[0] = 0xFF;
            SendBuff[1] = 0x00;
            SendBuff[2] = 0x00;
            SendBuff[3] = 0x00;     
            SendBuff[4] = (byte)(datalen);
            SendBuff[5] = 0xD4;
            SendBuff[6] = 0x40;
            SendBuff[7] = 0x01;

            for (index = 0; index <= datalen - 1 ; index++)
            {
                SendBuff[index + 8] = (byte)((int)(tempdata[index]));

                
            }

            SendLen = datalen + 8;
            RecvLen = 7;

            retCode = CardControl();
            if (retCode != ModWinsCard.SCARD_S_SUCCESS)
            {
                displayOut(2,retCode,"");
                return;
            }
        }
    

    
    //////////////////////////////////////////////////////////////////////
    }

}